src/coreclr/jit/morphblock.cpp

```diff
@@ -0,0 +1,1551 @@
+    GenTreeLclVarCommon* m_dstLclNode         = nullptr;
+    LclVarDsc*           m_dstVarDsc          = nullptr;
+    FieldSeqNode*        m_dstFldSeq          = nullptr;
+    unsigned             m_dstLclOffset       = 0;
+    bool                 m_dstUseLclFld       = false;
+    bool                 m_dstSingleLclVarAsg = false;
+    GenTree*             m_dstAddr            = nullptr;
+    ssize_t              m_dstAddOff          = 0;
+
+#if defined(DEBUG)
+    bool m_isLateArg = false;
+#endif // DEBUG
+
+    enum class BlockTransformation
+    {
+        Undefined,
+        FieldByField,
+        OneAsgBlock,
+        StructBlock,
+        SkipCallSrc,
+        Nop
+    };
+
+    BlockTransformation m_transformationDecision = BlockTransformation::Undefined;
+    GenTree*            m_result                 = nullptr;
+};
+
+//------------------------------------------------------------------------
+// MorphInitBlock: Morph a block initialization assignment tree.
+//
+// Arguments:
+//    comp - a compiler instance;
+//    tree - A GT_ASG tree that performs block initialization.
+//
+// Return Value:
+//    A possibly modified tree to perfom the initializetion.
+//
+// static
+GenTree* MorphInitBlockHelper::MorphInitBlock(Compiler* comp, GenTree* tree)
+{
+    const bool           initBlock = false;
+    MorphInitBlockHelper helper(comp, tree, initBlock);
+    return helper.Morph();
+}
+
+//------------------------------------------------------------------------
+// MorphInitBlockHelper: helper's constructor.
+//
+// Arguments:
+//    comp - a compiler instance;
+//    initBlock - true if this is init block op, false if it is a copy block;
+//    asg - GT_ASG node to morph.
+//
+// Notes:
+//    Most class members are initialized via in-class member initializers.
+//
+MorphInitBlockHelper::MorphInitBlockHelper(Compiler* comp, GenTree* asg, bool initBlock = true)
+    : m_comp(comp), m_initBlock(initBlock)
+{
+    assert(asg->OperIs(GT_ASG));
+#if defined(DEBUG)
+    if (m_initBlock)
+    {
+        assert(asg->OperIsInitBlkOp());
+    }
+    else
+    {
+        assert(asg->OperIsCopyBlkOp());
+    }
+#endif // DEBUG
+    m_asg = asg->AsOp();
+}
+
+//------------------------------------------------------------------------
+// Morph: transform the asg to a possible better form and changes its children
+//    to an appropriate form for later phases, for example, adds SIMD_INIT nodes
+//    or sets lvDoNotEnregister on locals.
+//
+// Return Value:
+//    A possibly modified tree to perfom the block operation.
+//
+// Notes:
+//    It is used for both init and copy block.
+//
+GenTree* MorphInitBlockHelper::Morph()
+{
+    JITDUMP("%s:\n", GetHelperName());
+
+    PrepareDst();
+    PrepareSrc();
+
+    INDEBUG(m_isLateArg = (m_asg->gtFlags & GTF_LATE_ARG) != 0);
+
+    TrySpecialCases();
+
+    if (m_transformationDecision == BlockTransformation::Undefined)
+    {
+        GenTree* oneAsgTree = m_comp->fgMorphOneAsgBlockOp(m_asg);
```

Comment:
I am planning to delete `fgMorphOneAsgBlockOp` later so it is why I did not bother to move it into this class.